postgres/gcppostgres: SIGSEGV Trying to Connect to CloudSQL PostgreSQL Instance from Cloud Run

### Describe the bug

Trying to connect to a CloudSQL PostgreSQL Instance in Cloud Run using v0.26.0 causes a SIGSEGV:
```
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x18 pc=0xc93612]

goroutine 47 [running]:
github.com/GoogleCloudPlatform/cloudsql-proxy/proxy/certs.(*RemoteCertSource).Local(0xc000723b60, {0xc0007381e0, 0x2a})
	/go/pkg/mod/github.com/!google!cloud!platform/cloudsql-proxy@v1.32.0/proxy/certs/certs.go:225 +0x252
github.com/GoogleCloudPlatform/cloudsql-proxy/proxy/proxy.(*Client).refreshCfg(0xc00014b080, {0xc0007381e0, 0x2a})
	/go/pkg/mod/github.com/!google!cloud!platform/cloudsql-proxy@v1.32.0/proxy/proxy/client.go:205 +0x1bb
github.com/GoogleCloudPlatform/cloudsql-proxy/proxy/proxy.(*Client).startRefresh.func1()
	/go/pkg/mod/github.com/!google!cloud!platform/cloudsql-proxy@v1.32.0/proxy/proxy/client.go:284 +0xad
created by github.com/GoogleCloudPlatform/cloudsql-proxy/proxy/proxy.(*Client).startRefresh
	/go/pkg/mod/github.com/!google!cloud!platform/cloudsql-proxy@v1.32.0/proxy/proxy/client.go:282 +0xcc
```
Switching back to v0.25.0 solves the issue

### To Reproduce
Trying to connect to a CloudSQL PostgreSQL Instance from a Cloud Run Instance with v0.26.0:
```go
package main

import (
	"fmt"
	"gocloud.dev/postgres"
	_ "gocloud.dev/postgres/gcppostgres"
	"context"
)

func main(){
	db, err := postgres.Open(context.Background(), "gcppostgres://postgres:postgres@my-project/us-east4/cloudsqlinstance/postgres")
	if err != nil {
		panic(err)
	}

	err = db.Ping()
	if err != nil {
		panic(err)
	}

	fmt.Println("Database successfully pinged.")
}
```



## Expected behavior

Connecting without SIGSEGV

### Version

v0.26.0

### Additional context

I suspect this is related: https://github.com/google/go-cloud/pull/3122 but I doubled checked and my CloudSQL instance has the cloudsql.iam_authentication turned on.

Before the SIGSEGV the following was logged:
```
2022/09/14 02:38:19 refreshing ephemeral certificate for instance my-project:us-east4:cloudsqlinstance
2022/09/14 02:38:19 Generated RSA key in 250.263502ms
```

--- a/gcp/cloudsql/cloudsql.go
+++ b/gcp/cloudsql/cloudsql.go
@@ -36,8 +36,8 @@ func NewCertSource(c *gcp.HTTPClient) *certs.RemoteCertSource {
 	return certs.NewCertSourceOpts(&c.Client, certs.RemoteOpts{})
 }
 
-// NewCertSourceIam creates a local certificate source, including Token source for token information used in
-//cert creation, that uses the given HTTP client. The client is assumed to make authenticated requests.
+// NewCertSourceWithIAM creates a local certificate source, including Token source for token information used in
+// cert creation, that uses the given HTTP client. The client is assumed to make authenticated requests.
 func NewCertSourceWithIAM(c *gcp.HTTPClient, t oauth2.TokenSource) *certs.RemoteCertSource {
 	return certs.NewCertSourceOpts(&c.Client, certs.RemoteOpts{EnableIAMLogin: true, TokenSource: t})
 }

--- a/gcp/cloudsql/cloudsql.go
+++ b/gcp/cloudsql/cloudsql.go
@@ -21,6 +21,7 @@ import (
 	"github.com/GoogleCloudPlatform/cloudsql-proxy/proxy/proxy"
 	"github.com/google/wire"
 	"gocloud.dev/gcp"
+	"golang.org/x/oauth2"
 )
 
 // CertSourceSet is a Wire provider set that binds a Cloud SQL proxy
@@ -31,6 +32,6 @@ var CertSourceSet = wire.NewSet(
 
 // NewCertSource creates a local certificate source that uses the given
 // HTTP client. The client is assumed to make authenticated requests.
-func NewCertSource(c *gcp.HTTPClient) *certs.RemoteCertSource {
-	return certs.NewCertSourceOpts(&c.Client, certs.RemoteOpts{})
+func NewCertSource(c *gcp.HTTPClient, t oauth2.TokenSource) *certs.RemoteCertSource {
+	return certs.NewCertSourceOpts(&c.Client, certs.RemoteOpts{EnableIAMLogin: true, TokenSource: t})
 }

--- a/gcp/cloudsql/cloudsql.go
+++ b/gcp/cloudsql/cloudsql.go
@@ -32,6 +32,12 @@ var CertSourceSet = wire.NewSet(
 
 // NewCertSource creates a local certificate source that uses the given
 // HTTP client. The client is assumed to make authenticated requests.
-func NewCertSource(c *gcp.HTTPClient, t oauth2.TokenSource) *certs.RemoteCertSource {
+func NewCertSource(c *gcp.HTTPClient) *certs.RemoteCertSource {
+	return certs.NewCertSourceOpts(&c.Client, certs.RemoteOpts{})
+}
+
+// NewCertSourceIam creates a local certificate source, including Token source for token information used in
+//cert creation, that uses the given HTTP client. The client is assumed to make authenticated requests.
+func NewCertSourceWithIAM(c *gcp.HTTPClient, t oauth2.TokenSource) *certs.RemoteCertSource {
 	return certs.NewCertSourceOpts(&c.Client, certs.RemoteOpts{EnableIAMLogin: true, TokenSource: t})
 }

--- a/postgres/gcppostgres/gcppostgres.go
+++ b/postgres/gcppostgres/gcppostgres.go
@@ -75,7 +75,7 @@ func (o *lazyCredsOpener) OpenPostgresURL(ctx context.Context, u *url.URL) (*sql
 			o.err = err
 			return
 		}
-		certSource := cloudsql.NewCertSource(client)
+		certSource := cloudsql.NewCertSource(client, creds.TokenSource)
 		o.opener = &URLOpener{CertSource: certSource}
 	})
 	if o.err != nil {

--- a/postgres/gcppostgres/gcppostgres.go
+++ b/postgres/gcppostgres/gcppostgres.go
@@ -75,7 +75,7 @@ func (o *lazyCredsOpener) OpenPostgresURL(ctx context.Context, u *url.URL) (*sql
 			o.err = err
 			return
 		}
-		certSource := cloudsql.NewCertSource(client, creds.TokenSource)
+		certSource := cloudsql.NewCertSourceWithIAM(client, creds.TokenSource)
 		o.opener = &URLOpener{CertSource: certSource}
 	})
 	if o.err != nil {

--- a/postgres/gcppostgres/gcppostgres_test.go
+++ b/postgres/gcppostgres/gcppostgres_test.go
@@ -41,8 +41,9 @@ func TestURLOpener(t *testing.T) {
 	username, _ := tfOut["username"].Value.(string)
 	password, _ := tfOut["password"].Value.(string)
 	databaseName, _ := tfOut["database"].Value.(string)
-	if project == "" || region == "" || instance == "" || username == "" || databaseName == "" {
-		t.Fatalf("Missing one or more required Terraform outputs; got project=%q region=%q instance=%q username=%q database=%q", project, region, instance, username, databaseName)
+	userEmail, _ := tfOut["user_email"].Value.(string)
+	if project == "" || region == "" || instance == "" || username == "" || databaseName == "" || userEmail == "" {
+		t.Fatalf("Missing one or more required Terraform outputs; got project=%q region=%q instance=%q username=%q database=%q userEmail=%q", project, region, instance, username, databaseName, userEmail)
 	}
 	tests := []struct {
 		name    string
@@ -50,7 +51,11 @@ func TestURLOpener(t *testing.T) {
 		wantErr bool
 	}{
 		{
-			name:   "Success",
+			name:   "SuccessIam",
+			urlstr: fmt.Sprintf("gcppostgres://%s@%s/%s/%s/%s", userEmail, project, region, instance, databaseName),
+		},
+		{
+			name:   "SuccessBuiltin",
 			urlstr: fmt.Sprintf("gcppostgres://%s:%s@%s/%s/%s/%s", username, password, project, region, instance, databaseName),
 		},
 		{

--- a/postgres/gcppostgres/main.tf
+++ b/postgres/gcppostgres/main.tf
@@ -42,8 +42,8 @@ variable "user_email" {
 }
 
 variable "region" {
-  default     = "europe-west1"
-  description = "GCP region to create database and storage in, for example 'europe-west1'. See https://cloud.google.com/compute/docs/regions-zones/ for valid values."
+  default     = "us-central1"
+  description = "GCP region to create database and storage in, for example 'us-central1'. See https://cloud.google.com/compute/docs/regions-zones/ for valid values."
 }
 
 locals {

--- a/postgres/gcppostgres/main.tf
+++ b/postgres/gcppostgres/main.tf
@@ -15,26 +15,35 @@
 # Harness for Cloud SQL Postgres tests.
 
 terraform {
-  required_version = "~>0.12"
+  required_version = ">= 1.1.0"
+  required_providers {
+    google = {
+      version = "4.40.0"
+    }
+    random = {
+      version = "3.4.3"
+    }
+  }
 }
 
 provider "google" {
-  version = "~> 2.5"
   project = var.project
-}
-
-provider "random" {
-  version = "~> 2.1"
+  region  = var.region
 }
 
 variable "project" {
   type        = string
-  description = "Project to set up."
+  description = "Project ID - Google Cloud project ID in which to create resources."
+}
+
+variable "user_email" {
+  type        = string
+  description = "User email address - Google identity to be used for testing IAM authentication."
 }
 
 variable "region" {
-  default     = "us-central1"
-  description = "GCP region to create database and storage in, for example 'us-central1'. See https://cloud.google.com/compute/docs/regions-zones/ for valid values."
+  default     = "europe-west1"
+  description = "GCP region to create database and storage in, for example 'europe-west1'. See https://cloud.google.com/compute/docs/regions-zones/ for valid values."
 }
 
 locals {
@@ -60,6 +69,18 @@ resource "random_id" "sql_instance" {
   byte_length = 12
 }
 
+resource "google_project_iam_member" "cloudsql_client" {
+  project = var.project
+  role    = "roles/cloudsql.client"
+  member  = "user:${var.user_email}"
+}
+
+resource "google_project_iam_member" "cloudsql_instanceUser" {
+  project = var.project
+  role    = "roles/cloudsql.instanceUser"
+  member  = "user:${var.user_email}"
+}
+
 resource "google_sql_database_instance" "main" {
   name             = local.sql_instance
   database_version = "POSTGRES_9_6"
@@ -69,6 +90,10 @@ resource "google_sql_database_instance" "main" {
   settings {
     tier      = "db-f1-micro"
     disk_size = 10 # GiB
+    database_flags {
+      name  = "cloudsql.iam_authentication"
+      value = "on"
+    }
   }
 
   depends_on = [
@@ -78,6 +103,7 @@ resource "google_sql_database_instance" "main" {
 }
 
 resource "google_sql_database" "main" {
+  project  = var.project
   name     = "testdb"
   instance = google_sql_database_instance.main.name
 }
@@ -94,9 +120,16 @@ resource "random_string" "db_password" {
 }
 
 resource "google_sql_user" "root" {
+  type     = "BUILT_IN"
   name     = "root"
   instance = google_sql_database_instance.main.name
   password = random_string.db_password.result
+}
+
+resource "google_sql_user" "user_account" {
+  type     = "CLOUD_IAM_USER"
+  name     = var.user_email
+  instance = google_sql_database_instance.main.name
 }
 
 output "project" {
@@ -130,3 +163,7 @@ output "database" {
   description = "The name of the database inside the Cloud SQL instance."
 }
 
+output "user_email" {
+  value       = var.user_email
+  description = "The email of a GCP service account used for testing connections."
+}

